File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/HistoryStore.swift
import Foundation
import VibeUI

/// Local, on-device dictation history.
///
/// Every dictation final is appended to
/// `~/Library/Application Support/VibeXASR/history.json` — a JSON array of
/// `{text, date}` (date = ISO-8601). File writes go through a private serial
/// `DispatchQueue` so concurrent appends can't corrupt the file; reads tolerate
/// a missing or corrupt file (returning an empty list). Nothing here ever
/// touches the network — the data stays on this device.
///
/// The in-memory `entries` (newest-first) is `@Published` so the History window
/// updates live. `HistoryItem` is the VibeUI-facing row type so the window can
/// render without importing this target.
@MainActor
final class HistoryStore: ObservableObject, HistoryBridge {

    static let shared = HistoryStore()

    /// Newest-first, for display.
    @Published private(set) var entries: [HistoryItem] = []

    /// Lifetime cumulative character count dictated. Persisted separately (in
    /// UserDefaults) so it SURVIVES clearing the history list — it only ever grows.
    @Published private(set) var lifetimeChars: Int = UserDefaults.standard.integer(forKey: lifetimeCharsKey) {
        didSet { UserDefaults.standard.set(lifetimeChars, forKey: Self.lifetimeCharsKey) }
    }
    private static let lifetimeCharsKey = "historyLifetimeChars"

    private let queue = DispatchQueue(label: "com.xasr.vibexasr.history")
    private let fileURL: URL

    private struct Record: Codable {
        var id: String?     // optional for old files; persisted so ids stay stable across reloads
        var text: String
        var date: Date
        var mode: String?   // "manual" | "oncall"; optional for old files
        var tags: [String]? // optional for old files
        var title: String?  // note title (整理成笔记); nil for normal entries
    }

    /// HistoryItem → on-disk Record (used by persist + export).
    private func record(_ e: HistoryItem) -> Record {
        Record(id: e.id.uuidString, text: e.text, date: e.date, mode: e.mode,
               tags: e.tags.isEmpty ? nil : e.tags, title: e.title)
    }

    init() {
        self.fileURL = ModelPaths.appSupportDir().appendingPathComponent("history.json")
        load()
    }

    // MARK: HistoryBridge (read API for the window)

    var historyItems: [HistoryItem] { entries }

    /// Append a final (called on each engine.onFinal). Ignores blank text.
    /// Seconds an ephemeral (history-disabled) record survives before self-destruct.
    static let ephemeralTTL: TimeInterval = 60

    /// Append a final. When `ephemeral` (history saving is OFF) the entry is kept in
    /// memory only for `ephemeralTTL` seconds (shown with a countdown) then removed,
    /// and never written to disk — a grace buffer so a long unsaved dictation isn't
    /// lost the instant it ends.
    func append(_ text: String, mode: String = "manual", ephemeral: Bool = false) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        let expires = ephemeral ? Date().addingTimeInterval(Self.ephemeralTTL) : nil
        // Store the trimmed text — stray leading whitespace (common in ASR output)
        // would indent the row vs its timestamp.
        let item = HistoryItem(id: UUID(), text: trimmed, date: Date(), mode: mode, expiresAt: expires)
        entries.insert(item, at: 0)               // newest first in memory
        lifetimeChars += trimmed.count            // cumulative; survives clearing
        if let expires {
            let id = item.id
            let delay = expires.timeIntervalSinceNow
            DispatchQueue.main.asyncAfter(deadline: .now() + max(0, delay)) { [weak self] in
                self?.entries.removeAll { $0.id == id }   // self-destruct (never persisted)
            }
        } else {
            persistAsync()
        }
    }

    /// Delete a single entry by id.
    func delete(id: UUID) {
        entries.removeAll { $0.id == id }
        persistAsync()
    }

    /// Edit an entry's text in place (issue #6). Keeps id + date; does NOT change
    /// lifetimeChars (which counts what was dictated, not the current edited text).
    func update(id: UUID, text: String) {
        let cur = entries.first { $0.id == id }
        update(id: id, text: text, title: cur?.title, tags: cur?.tags ?? [])
    }

    /// Rich edit (inline editor): text + note title + tags. Keeps id + date + mode.
    func update(id: UUID, text: String, title: String?, tags: [String]) {
        guard let idx = entries.firstIndex(where: { $0.id == id }) else { return }
        let old = entries[idx]
        entries[idx] = HistoryItem(id: old.id, text: text, date: old.date, mode: old.mode,
                                   expiresAt: old.expiresAt, tags: tags, title: title)
        persistAsync()
    }

    /// Merge entries → one. Ascending-date join; newest entry is the anchor (kept),
    /// others removed. asNote → newline-join + note title; else direct concat. Tags
    /// unioned; mode = "oncall" iff every merged entry was oncall.
    func merge(ids: [UUID], asNote: Bool, title: String?) {
        let set = Set(ids)
        // entries is newest-first → index 0 is the newest (top of the list).
        let order = Dictionary(entries.enumerated().map { ($1.id, $0) }, uniquingKeysWith: { a, _ in a })
        // Merge in DISPLAY order: the entry shown ABOVE (newer) comes first, the one
        // BELOW (older) after — matching what the user sees top→bottom. Stable
        // tiebreak on list position so same-timestamp fragments never scramble.
        let chosen = entries.filter { set.contains($0.id) }.sorted { a, b in
            a.date != b.date ? a.date > b.date : (order[a.id] ?? 0) < (order[b.id] ?? 0)
        }
        guard chosen.count >= (asNote ? 1 : 2) else { return }
        let text = chosen.map(\.text).joined(separator: asNote ? "\n" : "")
        guard let anchor = chosen.first else { return }   // top (newest) entry = anchor → "合并到上面的时间"
        var mergedTags: [String] = []
        for e in chosen { for t in e.tags where !mergedTags.contains(t) { mergedTags.append(t) } }
        let allOnCall = chosen.allSatisfy { $0.mode == "oncall" }
        entries = entries.compactMap { e in
            if e.id == anchor.id {
                return HistoryItem(id: e.id, text: text, date: e.date,
                                   mode: allOnCall ? "oncall" : "manual",
                                   expiresAt: e.expiresAt, tags: mergedTags,
                                   title: asNote ? (title ?? "整理笔记") : nil)
            }
            return set.contains(e.id) ? nil : e
        }
        persistAsync()
    }

    /// Union `tag` into each id.
    func applyTag(ids: [UUID], tag: String) {
        let set = Set(ids)
        entries = entries.map { e in
            guard set.contains(e.id), !e.tags.contains(tag) else { return e }
            return HistoryItem(id: e.id, text: e.text, date: e.date, mode: e.mode,
                               expiresAt: e.expiresAt, tags: e.tags + [tag], title: e.title)
        }
        persistAsync()
    }

    /// Replace an entry's tags wholesale.
    func setTags(id: UUID, tags: [String]) {
        guard let idx = entries.firstIndex(where: { $0.id == id }) else { return }
        let old = entries[idx]
        entries[idx] = HistoryItem(id: old.id, text: old.text, date: old.date, mode: old.mode,
                                   expiresAt: old.expiresAt, tags: tags, title: old.title)
        persistAsync()
    }

    /// Insert a blank manual entry at now; returns its id for immediate editing.
    @discardableResult func addEntry() -> UUID {
        let item = HistoryItem(id: UUID(), text: "", date: Date(), mode: "manual")
        entries.insert(item, at: 0)
        persistAsync()
        return item.id
    }

    /// Replace the whole list (undo restore). lifetimeChars is unaffected.
    func replaceAll(_ items: [HistoryItem]) {
        entries = items
        persistAsync()
    }

    /// Clear all history.
    func clearAll() {
        entries.removeAll()
        lifetimeChars = 0          // clearing records also resets the cumulative stats
        persistAsync()
    }

    // MARK: export (issue #5)

    /// JSON representation of all history (chronological, oldest-first), ready to
    /// write to a user-chosen file via NSSavePanel. The NSSavePanel itself lives in
    /// the AppKit-capable HistoryView.
    func exportJSONData() -> Data {
        let records = entries.reversed().filter { $0.expiresAt == nil }.map(record)
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return (try? encoder.encode(records)) ?? Data("[]".utf8)
    }

    /// Plain-text representation: one entry per block, "<ISO date>\t<text>"
    /// chronological (oldest-first).
    func exportPlainText() -> String {
        let iso = ISO8601DateFormatter()
        return entries.reversed()
            .map { "\(iso.string(from: $0.date))\t\($0.text)" }
            .joined(separator: "\n")
    }

    // MARK: persistence

    /// Load synchronously at startup; tolerate missing/corrupt file.
    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        guard let records = try? decoder.decode([Record].self, from: data) else {
            // Corrupt file — start fresh (don't crash). The next write overwrites it.
            return
        }
        // Stored oldest-first; present newest-first.
        entries = records.reversed().map {
            HistoryItem(id: $0.id.flatMap(UUID.init) ?? UUID(),
                        text: $0.text, date: $0.date, mode: $0.mode ?? "manual",
                        tags: $0.tags ?? [], title: $0.title)
        }
    }

    /// Snapshot the current entries and write them off the main thread.
    private func persistAsync() {
        // Store oldest-first (chronological) so external readers see natural order.
        let records = entries.reversed().filter { $0.expiresAt == nil }.map(record)
        let url = fileURL
        queue.async {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            encoder.outputFormatting = [.prettyPrinted]
            guard let data = try? encoder.encode(records) else { return }
            // Atomic write so a crash mid-write can't truncate the file.
            try? data.write(to: url, options: .atomic)
        }
    }
}

File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Storage/HistoryStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace VibeXASR.Windows.Storage;

/// <summary>
/// One recognized utterance. Mirrors the macOS <c>HistoryItem</c>: stable id, a
/// timestamp, the text, the dictation mode that produced it ("paste"/"type"/"oncall"),
/// and an optional <see cref="ExpiresAt"/> for ephemeral records kept only ~60 s when
/// "save history" is off.
/// </summary>
public sealed class HistoryEntry
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public DateTimeOffset Timestamp { get; set; } = DateTimeOffset.Now;
    public string Text { get; set; } = string.Empty;
    public string Mode { get; set; } = "paste";
    public DateTimeOffset? ExpiresAt { get; set; }

    // v1.4.0 history workspace: per-entry tags + an optional title (for clustered/merged blocks).
    public List<string> Tags { get; set; } = new();
    public string? Title { get; set; }
}

/// <summary>On-disk shape: cumulative stats + the entry list (object form).</summary>
internal sealed class HistoryFile
{
    public long LifetimeChars { get; set; }
    public List<HistoryEntry> Entries { get; set; } = new();
}

/// <summary>
/// Local dictation history persisted as JSON in %APPDATA%/VibeXASR/history.json.
/// Loaded fully into memory (small); each mutation rewrites the whole file
/// (atomic write-then-rename). Tracks a cumulative character count that survives
/// per-row deletion (only "Clear all" resets it), matching the macOS History stats.
/// </summary>
public sealed class HistoryStore
{
    private readonly object _gate = new();
    private readonly List<HistoryEntry> _entries = new();
    private long _lifetimeChars;

    private static readonly JsonSerializerOptions JsonOpts = new() { WriteIndented = true };

    public static string FilePath => Path.Combine(AppPaths.DataDir, "history.json");

    /// <summary>Cumulative characters dictated (survives row deletion; reset by ClearAll).</summary>
    public long LifetimeChars { get { lock (_gate) return _lifetimeChars; } }

    public HistoryStore() => Reload();

    public void Reload()
    {
        lock (_gate)
        {
            _entries.Clear();
            _lifetimeChars = 0;
            try
            {
                if (File.Exists(FilePath))
                {
                    var json = File.ReadAllText(FilePath);
                    var file = TryParse(json);
                    if (file is not null)
                    {
                        _entries.AddRange(file.Entries);
                        _lifetimeChars = file.LifetimeChars;
                    }
                }
            }
            catch { /* corrupt history -> start empty */ }
            PruneExpired();
        }
    }

    /// <summary>Parse the object form; fall back to a bare array (legacy skeleton format).</summary>
    private static HistoryFile? TryParse(string json)
    {
        try { return JsonSerializer.Deserialize<HistoryFile>(json, JsonOpts); }
        catch { }
        try
        {
            var arr = JsonSerializer.Deserialize<List<HistoryEntry>>(json, JsonOpts);
            if (arr is not null)
                return new HistoryFile { Entries = arr, LifetimeChars = arr.Sum(e => (long)e.Text.Length) };
        }
        catch { }
        return null;
    }

    /// <summary>
    /// Append a finalized utterance. <paramref name="ephemeral"/> records (history off)
    /// get a 60 s <see cref="HistoryEntry.ExpiresAt"/> and are pruned automatically.
    /// Always advances the cumulative <see cref="LifetimeChars"/>.
    /// </summary>
    public HistoryEntry? Append(string text, string mode = "paste", bool ephemeral = false)
    {
        if (string.IsNullOrWhiteSpace(text)) return null;
        var trimmed = text.Trim();
        lock (_gate)
        {
            var entry = new HistoryEntry
            {
                Text = trimmed,
                Mode = mode,
                ExpiresAt = ephemeral ? DateTimeOffset.Now.AddSeconds(60) : null,
            };
            _entries.Add(entry);
            _lifetimeChars += trimmed.Length;
            Persist();
            return entry;
        }
    }

    /// <summary>Most-recent-first snapshot, expired records already dropped.</summary>
    public IReadOnlyList<HistoryEntry> List()
    {
        lock (_gate)
        {
            PruneExpired();
            return _entries.OrderByDescending(e => e.Timestamp).ToList();
        }
    }

    public void Delete(Guid id)
    {
        lock (_gate)
        {
            if (_entries.RemoveAll(e => e.Id == id) > 0) Persist();
        }
    }

    public void Update(Guid id, string text)
    {
        lock (_gate)
        {
            var e = _entries.FirstOrDefault(x => x.Id == id);
            if (e is not null) { e.Text = text.Trim(); Persist(); }
        }
    }

    // ===== v1.4.0 history-workspace operations =====

    /// <summary>Rich edit (inline editor): text + optional note title + tags. Keeps id/date/mode.</summary>
    public void Update(Guid id, string text, string? title, List<string> tags)
    {
        lock (_gate)
        {
            var e = _entries.FirstOrDefault(x => x.Id == id);
            if (e is null) return;
            e.Text = text.Trim();
            e.Title = string.IsNullOrWhiteSpace(title) ? null : title.Trim();
            e.Tags = tags?.ToList() ?? new();
            Persist();
        }
    }

    /// <summary>Merge entries → one. Newest entry is the anchor (kept), others removed. Tags unioned;
    /// mode="oncall" iff every merged entry was oncall, else "manual". asNote → newline-join + note
    /// title; else direct concat. Join is in DISPLAY (newest-first) order, matching macOS.</summary>
    public void Merge(IReadOnlyList<Guid> ids, bool asNote, string? title)
    {
        lock (_gate)
        {
            var set = new HashSet<Guid>(ids);
            var displayIdx = _entries.OrderByDescending(e => e.Timestamp).Select((e, i) => (e.Id, i))
                                     .ToDictionary(x => x.Id, x => x.i);
            var chosen = _entries.Where(e => set.Contains(e.Id))
                                 .OrderByDescending(e => e.Timestamp)
                                 .ThenBy(e => displayIdx.TryGetValue(e.Id, out var i) ? i : 0)
                                 .ToList();
            if (chosen.Count < (asNote ? 1 : 2)) return;
            var anchor = chosen[0];
            anchor.Text = string.Join(asNote ? "\n" : "", chosen.Select(e => e.Text));
            var mergedTags = new List<string>();
            foreach (var e in chosen) foreach (var t in e.Tags) if (!mergedTags.Contains(t)) mergedTags.Add(t);
            anchor.Tags = mergedTags;
            anchor.Mode = chosen.All(e => e.Mode == "oncall") ? "oncall" : "manual";
            anchor.Title = asNote ? (string.IsNullOrWhiteSpace(title) ? "整理笔记" : title.Trim()) : null;
            set.Remove(anchor.Id);
            _entries.RemoveAll(e => set.Contains(e.Id));
            Persist();
        }
    }

    /// <summary>Add a tag to several entries (skips ones that already have it).</summary>
    public void ApplyTag(IReadOnlyList<Guid> ids, string tag)
    {
        if (string.IsNullOrWhiteSpace(tag)) return;
        tag = tag.Trim();
        lock (_gate)
        {
            var set = new HashSet<Guid>(ids);
            bool changed = false;
            foreach (var e in _entries)
                if (set.Contains(e.Id) && !e.Tags.Contains(tag)) { e.Tags.Add(tag); changed = true; }
            if (changed) Persist();
        }
    }

    /// <summary>Replace one entry's tags wholesale.</summary>
    public void SetTags(Guid id, List<string> tags)
    {
        lock (_gate)
        {
            var e = _entries.FirstOrDefault(x => x.Id == id);
            if (e is null) return;
            e.Tags = tags?.ToList() ?? new();
            Persist();
        }
    }

    /// <summary>Insert a new empty "manual" note at the top; returns its id (for inline editing).</summary>
    public Guid AddEntry()
    {
        lock (_gate)
        {
            var e = new HistoryEntry { Text = "", Mode = "manual" };
            _entries.Add(e);
            Persist();
            return e.Id;
        }
    }

    /// <summary>Deep-copy snapshot of all entries (for the workspace undo stack).</summary>
    public List<HistoryEntry> Snapshot()
    {
        lock (_gate) return _entries.Select(Clone).ToList();
    }

    /// <summary>Restore a previous snapshot wholesale (undo).</summary>
    public void Restore(List<HistoryEntry> snapshot)
    {
        lock (_gate)
        {
            _entries.Clear();
            _entries.AddRange(snapshot.Select(Clone));
            Persist();
        }
    }

    private static HistoryEntry Clone(HistoryEntry e) => new()
    {
        Id = e.Id, Timestamp = e.Timestamp, Text = e.Text, Mode = e.Mode,
        ExpiresAt = e.ExpiresAt, Tags = new List<string>(e.Tags), Title = e.Title,
    };

    /// <summary>Clear all rows AND reset the cumulative stats (destructive).</summary>
    public void ClearAll()
    {
        lock (_gate)
        {
            _entries.Clear();
            _lifetimeChars = 0;
            Persist();
        }
    }

    /// <summary>Drop any records whose ephemeral lifetime elapsed. Caller holds the lock.</summary>
    private void PruneExpired()
    {
        var now = DateTimeOffset.Now;
        _entries.RemoveAll(e => e.ExpiresAt is { } exp && exp <= now);
    }

    private void Persist()
    {
        Directory.CreateDirectory(AppPaths.DataDir);
        var file = new HistoryFile { LifetimeChars = _lifetimeChars, Entries = _entries };
        var json = JsonSerializer.Serialize(file, JsonOpts);
        var tmp = FilePath + ".tmp";
        File.WriteAllText(tmp, json);
        File.Copy(tmp, FilePath, overwrite: true);
        File.Delete(tmp);
    }
}
